### PR TITLE
Add -alsaname command line flag to allow Pd to use unique Alsa midi port names

### PR DESCRIPTION
**Pull request by tstex**
Fix for: https://github.com/pure-data/pure-data/issues/2462

### CODE REVIEW

**Comment by umlaeute** [comment]
i wonder whether we shouldn't just re-use the `-jackname` for this.

obviously the name of the flag is sub-optimal (there's no JACK), but maybe the solution is to just rename the flag to `-devicename` (keeping `-jackname` as an alias).

**Comment by tstex** [comment]
I agree that is a good idea. I believe I tried that initially, but ran into problems. Let me try again tonight after work to verify.

**Comment by reduzent** [comment]
The flag `-jackname` does not only set the jack client name, but it implicitly switches the audio backend to JACK. I think a new flag like `-devicename` that only sets the name (for whatever audio or MIDI backend) would make a lot of sense and would allow a better behavior without breaking backwards compatibility.

**Comment by tstex** [comment]
That kind of makes -jackname redundant in some situations yes? How do we resolve situations where the user uses both flags - but with different values?

I'll look at re-using the jackname value shortly. Are there any other use cases for the more generic -devicename on other platforms etc?

As an aside, I don't think jackname should set or switch the backend, but backwards compatibility seems to be highly prized.

**Comment by reduzent** [comment]
> As an aside, I don't think jackname should set or switch the backend, but backwards compatibility seems to be highly prized.

I absolutely agree on both points.

**Comment by umlaeute** [comment]
> Hmm. Ok. I remember why I didn't re-use -jackname now. Linkage hell. Makefile changes: (USEAPI_JACK in USEAPI_ALSA code). Much harder (impossible?) to do cleanly from a code point of view.

Nah, much simpler. 
`-devicename` sets an API independent name. As such, it *must not* be depending on any preprocessor define like `USEAPI_JACK` or the like. 
It will simply set a global symbol, that can then be used by any API that requires it (Jack, alsa-midi, futuresound,...)

> As an aside, I don't think jackname should set or switch the backend, but backwards compatibility seems to be highly prized.

I agree with the former.
Personally I could happily  live with `-jackname` not setting the backend. 
Iirc, the current behaviour is not documented (at least I hope so. There's a tendency to document every voidness, turning arbitrary (and not well-thought-of) decisions into canonical and expected behaviour), so we could just regard it as a bug. 
It's easy (and backward compatible) enough to just add an additional "-jack" flag. 
I *would* keep the compost flag name ("-jackname") around, so that startup scripts don't start to fail.

**Comment by reduzent** [comment]
I fully agree. I also rather would have `-jackname` not implicitly set the back-end. What I meant to say: I fear backwards compatibility concerns (I don't agree with them).

File: src/s_midi_alsa.c [hunk 1]
```diff
@@ -103,9 +104,13 @@ void sys_alsa_do_open_midi(int nmidiin, int *midiinvec,
         if (port < 0) goto error;
     }
 
+    if (!alsa_client_name || !strlen(alsa_client_name))
+        alsa_set_client_name("Pure Data");
+    post("Alsa MIDI client name:%s", alsa_client_name);
```
Comment (umlaeute) [inline src/s_midi_alsa.c:109]: this debugging output shouldn't be here

**Review by umlaeute — CHANGES_REQUESTED** [review]
instead of writing to an API-specific variable, i would instead use a global (but private) `sys_devicename[MAXPDSTRING]`  that get's initialized to `"Pure Data"` (or `"pure_data"`, but I'd actually prefer the former even if the rename breaks my qjackctl patchbays).

than JACK, or ALSA or whoever wants can use this name (as is, without having to check for emptiness).

**Comment by umlaeute** [comment]
why has this PR been closed?

**Comment by tstex** [comment]
I cleaned up my git history - which inadvertently closed the PR. I will
make a new one. I'm still working on it. There will be another one
involving autoconf probably. CFLAGS includes both -o0 and -o2 at present,
so the things I'm trying to look at in GDB have been optimized away. It's
been quite the side track. :D

But the pd code is written. I just need to debug this last thing....
